### DISCUSSION TITLE
Help resolve Issue re-directing to the dashboard

### DISCUSSION THREAD

**Discussion by PtradeLLC** [nhost/nhost#2271] [Q&A] [answered]
Hello, started building with Nhost, and I have questions :)

I was wondering why this code is not running the `if block` ,  instead it's logging the `else` statement in the console. However, and surprisingly it also logs users in (which means there is session created) , but doesn't redirect. Please help:

```
// SignIn form.
import React, { useState } from 'react';
import { useRouter } from 'next/router';
import { useSignInEmailPassword } from '@nhost/nextjs';
import Link from 'next/link';
import Loading from './Loading';
import PasswordReset from './PassReset';

const SignIn = () => {
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [open, setOpen] = useState(false);

    const router = useRouter();

    const { signInEmailPassword, isLoading, isSuccess, needsEmailVerification, isError, error } = useSignInEmailPassword();

    const handleOnSubmit = async (e) => {
        e.preventDefault();
        try {
            const data = await signInEmailPassword(email, password);

            if (isSuccess) {
                router.push('/dashboard');
                console.log("yay!");
                return null

            } else {
                console.log("There is an error", error);  // 'error' returns `null`
            }
        } catch (error) {
            console.error(error.message);
            setErrors(error.message);
        }
    };

    const setErrors = (error) => {
        return error.message
    };

    const disableForm = isLoading || needsEmailVerification;

return( {...rest of the code})


}
```
Please help

**Comment by dbarrosop** [comment]
I am not a frontend developer myself but I **think** the issue might be that the line:

```
const { signInEmailPassword, isLoading, isSuccess, needsEmailVerification, isError, error } = useSignInEmailPassword();
```

should be inside the method:

```
const handleOnSubmit = async (e) => {
```

You can also find a working example you can use as reference here:

https://github.com/nhost/nhost/blob/main/examples/react-apollo/src/sign-in/email-password.tsx

Which is deployed on this url:

https://react-apollo.example.nhost.io/sign-in

Let us know if my suggestion works or if you manage to figure it out with the example provided.

**Reply by PtradeLLC** [reply]
Hello, and thanks for your prompt response. I have the `useSignInEmailPassword()` hook in the global state of the `Component` cos a React Hook has to be called inside the functional `Component` - defining the hook anywhere else throws an error. I'll review the resources you included in your response to see if it helps. Thanks

**Comment by onehassan** [comment, marked as answer]
If you want to handle the sign-in redirect inside the `handleOnSubmit` you use the `isSuccess` and `error` from the data you get from `signInEmailPassword`. So I would suggest you try this approach:

```javascript
    const { signInEmailPassword, isLoading } = useSignInEmailPassword();

    const handleOnSubmit = async (e) => {
        e.preventDefault();


        const { isSuccess, error } = await signInEmailPassword(email, password);

        if (isSuccess) {
            router.push('/dashboard');
        } else {
            console.log("There is an error", error);
            setErrors(error.message)
        }
    };

```

The `isSuccess` and `error` returned from `useSignInEmailPassword` will have the correct values on the next re-render of the `SignIn` form that's why you end up with `isSuccess` is false right after the `signInEmailPassword`

Hope this helps clear up the confusion. If you have any more questions let us know.

**Reply by PtradeLLC** [reply]
Thanks @onehassan , That fixed the issue....Nice!
However, I just posted another issue (https://github.com/nhost/nhost/discussions/2275)   :( ... Maybe you can help with that, you seem to have the Midas touch on my Nhost issues. Thanks in advance, and more thanks for the above solution.

Regards,
